apps/grader-host/src/index.ts

```diff
@@ -188,24 +207,80 @@ async function handleJob(job: GradingJobMessage) {
   try {
     const receivedTime = await timeReporter.reportReceivedTime(job.jobId);
 
-    const initResults = await Promise.all([
+    const initResults = await Promise.allSettled([
       reportReceived(context, receivedTime),
+      // It's possible that this could fail for normal reasons, e.g. a Docker
+      // image not existing in our cache. We'll allow this error to bubble up
+      // but we won't report it to Sentry.
       initDocker(context),
-      initFiles(context),
+      initFiles(context).catch((err) => {
+        Sentry.captureException(err);
+        throw err;
+      }),
     ]);
 
-    const results = await runJob(context, receivedTime, initResults[2].tempDir);
+    const results = await run(async () => {
+      const initSucceeded = initResults.every((result) => result.status === 'fulfilled');
+      if (initSucceeded && initResults[2].status === 'fulfilled') {
```

Comment:
If `initSucceeded` then we don't need the extra test.
```suggestion
      if (initSucceeded) {
```

Reply:
Yes we do: TypeScript can't infer the type of `initResults[2]` based on `initResults.every(...)`.

<img width="781" alt="Screenshot 2024-09-10 at 12 00 53" src="https://github.com/user-attachments/assets/582ea3c0-0a7f-43f8-9c32-4d4da6bbf646">

Reply:
Ah, fair, good to know I'm still smarter than TS in at least some circumstances.

Reply:
It's definitely nice to have a bit less of the magic `[2]` indexing into the results array 😅

I was randomly reading the TC39 meeting minutes last night (trying to understand when we might get `Temporal` in v8) and I saw that they are also planning on adding a `Promise.allKeyed()` that will allow:
```
const initResults = Promise.allKeyed({
  'received': reportReceived(context, receivedTime),
  'docker': initDocker(context),
  'files': initFiles(context).catch((err) => {
    Sentry.captureException(err);
    throw err;
  }),
});

const tempDir = initResults.files.status === 'fulfilled' ? initResults.files.value : undefined;
```